src/audio/selector/selector.c

```diff
@@ -83,6 +91,140 @@ static struct comp_dev *selector_new(const struct comp_driver *drv,
+
+	dev = comp_alloc(drv, sizeof(*dev));
+	if (!dev)
+		return NULL;
+	dev->ipc_config = *config;
+
+	cd = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(*cd));
+	if (!cd) {
+		rfree(dev);
+		return NULL;
+	}
+
+	comp_set_drvdata(dev, cd);
+
+	ret = memcpy_s(&cd->md, sizeof(cd->md), ipc_process, sizeof(*ipc_process));
+	assert(!ret);
+
+	build_config(cd);
+
+	dev->state = COMP_STATE_READY;
+	return dev;
+}
+
+static void set_selector_params(struct comp_dev *dev,
+				struct sof_ipc_stream_params *params)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	struct comp_buffer __sparse_cache *source;
+	struct comp_buffer __sparse_cache *source_b;
+	struct list_item *sink_list;
+	int i;
+
+	if (dev->direction == SOF_IPC_STREAM_PLAYBACK)
+		params->channels = cd->config.in_channels_count;
+	else
+		params->channels = cd->config.out_channels_count;
+
+	params->rate = cd->md.base_cfg.audio_fmt.sampling_frequency;
+	params->frame_fmt = cd->source_format;
+
+	for (i = 0; i < SOF_IPC_MAX_CHANNELS; i++)
+		params->chmap[i] = (cd->md.output_format.ch_map >> i * 4) & 0xf;
+
+	/* update each sink format */
+	list_for_item(sink_list, &dev->bsink_list) {
+		struct comp_buffer *sink_b =
+			container_of(sink_list, struct comp_buffer, source_list);
+		struct comp_buffer __sparse_cache *sink = buffer_acquire(sink_b);
+		struct ipc4_audio_format *out_fmt = &cd->md.output_format;
+
+		sink->stream.channels = params->channels;
+		sink->stream.rate = params->rate;
+		audio_stream_fmt_conversion(out_fmt->depth,
+					    out_fmt->valid_bit_depth,
+					    &sink->stream.frame_fmt,
+					    &sink->stream.valid_sample_fmt,
+					    out_fmt->s_type);
+
+		sink->buffer_fmt = out_fmt->interleaving_style;
+
+		for (i = 0; i < SOF_IPC_MAX_CHANNELS; i++)
+			sink->chmap[i] = (out_fmt->ch_map >> i * 4) & 0xf;
+
+		sink->hw_params_configured = true;
+		buffer_release(sink);
+	}
+
+	/* update the source format
+	 * used only for rare cases where two pipelines are connected by a shared
+	 * buffer and 2 copiers, this will set source format only for shared buffers
+	 * for a short time when the second pipeline already started
+	 * and the first one is not ready yet along with sink buffers params
+	 */
+	source_b = list_first_item(&dev->bsource_list, struct comp_buffer, sink_list);
+	source = buffer_acquire(source_b);
+	if (!source->hw_params_configured) {
+		struct ipc4_audio_format *in_fmt;
+
+		in_fmt = &cd->md.base_cfg.audio_fmt;
+		source->stream.channels = in_fmt->channels_count;
+		source->stream.rate = in_fmt->sampling_frequency;
+		audio_stream_fmt_conversion(in_fmt->depth,
+					    in_fmt->valid_bit_depth,
+					    &source->stream.frame_fmt,
+					    &source->stream.valid_sample_fmt,
+					    in_fmt->s_type);
+
+		source->buffer_fmt = in_fmt->interleaving_style;
+
+		for (i = 0; i < SOF_IPC_MAX_CHANNELS; i++)
+			source->chmap[i] = (in_fmt->ch_map >> i * 4) & 0xf;
+
+		source->hw_params_configured = true;
+		buffer_release(source);
```

Comment:
this must be outside of `if` now of course too.

Reply:
thanks